[RFC] Add a Mutex trait (#377)
rust-embedded/wg: rfcs/0377-mutex-trait.md

> Should the `Mutex` trait not be in the [`embedded-hal`]?

It is our strong recommendation to **not** place it in the [`embedded-hal`] crate.
If there are breaking changes in the [`embedded-hal`] it can be so that competing implementations of the Mutex trait starts to exist which could bifurcate the ecosystem based on version, plus that the Mutex trait will be tied to HAL releases.

# How We Teach This
[how-we-teach-this]: #how-we-teach-this

There will be 2 reference implementations after the acceptance of this RFC. One direct implementation based on critical sections in `cortex-m` and `risc-v` crates (that now reexport the [bare-metal mutex]) as the mutex will now be implemented in these crates and the reexport removed. And another in the [RTFM framework] which works together with procedural macros to give correct by construction resource handling and locks.

# Drawbacks
[drawbacks]: #drawbacks

As of writing this RFC I see no drawbacks.

# Alternatives
[alternatives]: #alternatives

The alternatives, taking `&self` and returning a `Result<...>` are discussed in [Detailed design](#detailed-design).

> One can also further follow the [Rust standard library `Mutex`] and its RAII pattern, however this should not be persued due to these causes:

Reviewer: ```suggestion
One can also further follow the [Rust standard library `Mutex`] and its RAII pattern, however this should not be pursued due to these causes:
```

> Should the `Mutex` trait not be in the [`embedded-hal`]?

It is our strong recommendation to **not** place it in the [`embedded-hal`] crate.
If there are breaking changes in the [`embedded-hal`] it can be so that competing implementations of the Mutex trait starts to exist which could bifurcate the ecosystem based on version, plus that the Mutex trait will be tied to HAL releases.

# How We Teach This
[how-we-teach-this]: #how-we-teach-this

There will be 2 reference implementations after the acceptance of this RFC. One direct implementation based on critical sections in `cortex-m` and `risc-v` crates (that now reexport the [bare-metal mutex]) as the mutex will now be implemented in these crates and the reexport removed. And another in the [RTFM framework] which works together with procedural macros to give correct by construction resource handling and locks.

# Drawbacks
[drawbacks]: #drawbacks

As of writing this RFC I see no drawbacks.

# Alternatives
[alternatives]: #alternatives

The alternatives, taking `&self` and returning a `Result<...>` are discussed in [Detailed design](#detailed-design).

> One can also futher follow the [Rust standard library `Mutex`] and its RAII pattern, however this should not be persued dues to these causes:

Reviewer: ```suggestion
One can also further follow the [Rust standard library `Mutex`] and its RAII pattern, however this should not be pursued dues to these causes:
```

1. The RAII version can be used to implement the closure and hence the closure can be implemented with an `std`-like mutex in the background.
2. The closure is explicit where it starts and ends, no need to wait for the destruction or add extra `{ ... }` around code to get predictable unlocks.
3. Makes code using locks more stable under refactoring, as small changes can drastically change the scope of a lock when using RAII.
4. A RAII based mutex can be `mem::forget` or `mem::swap` and all guarantees are out, the closure based mutex cannot be circumvented easily.

> 5. The RAII based mutex can be circumvented easily as follows:

Reviewer: ```suggestion
  4.1. The RAII based mutex can be circumvented easily as follows:
```

    fn lock<R>(&mut self, f: impl FnOnce(&mut Self::Data) -> R) -> R;
}
```

The design is such that the data protected by the mutex is only accessible within the closure provided to the `lock` method, moreover the closure shall run within a critical section such that it is guaranteed that no-one else can access the same data at the same time. The critical section can for example be:

* Disabling interrupts (e.g. `cortex-m` / `risc-v` critical sections)
* Analyzing the program and set BASEPRI in Cortex-M devices (Stack Resource Policy/RTFM way)
* Analyzing the program and set interrupt masks correctly
* etc

### Design decisions and compatibility

####  The [Rust standard library `Mutex`] uses `&self` for the `lock` method, why choose `&mut self`?

This is the **most important** part of this RFC proposal. Let's assume that the implementation was `&self`, what would this mean? This would allow the following code to compile, which is unsound by construction:

```rust
fn unsound(mtx: &impl core_mutex::Mutex<Data = i32>) {
    mtx.lock(|data| {
        *data += 1;

        mtx.lock(|data| *data += 1);
    });
}
```

> This example allows for a mutex to lock itself within a lock of itself, which is **unsound by construction** and creates an immediate deadlock - while if implementing it with `&mut self` the following would happen:

Reviewer 2: That's not unsound, it should deadlock, which is sound.

```rust
struct Mutex<T> {
    data: T,
}

impl<T> Mutex<T> {
    pub const fn new(data: T) -> Self {
        Self { data }
    }

    fn access(&self, _cs: &CriticalSection) -> &T {
        &self.data
    }
}

unsafe impl<'a, T> core_mutex::Mutex for &'a Mutex<RefCell<T>> {
    type Data = T;

    fn lock<R>(&mut self, f: impl FnOnce(&mut Self::Data) -> R) -> R {
        cortex_m::interrupt::free(|cs| f(&mut *self.access(cs).borrow_mut()))
    }
}
```

Which allows for the following program to exist, hence **not breaking compatibility** with existing pattern:

```rust
static MUT: Mutex<RefCell<i32>> = Mutex::new(RefCell::new(0));

#[entry]
fn init() -> ! {
    let mut r = &MUT; // Note that r is mutable

    // Locking!!
    r.lock(|data| *data += 1);

    loop {}
}

#[interrupt]
fn SWI0_EGU0() {
    let mut r = &MUT; // Note that r is mutable

    // Locking!!
    r.lock(|data| *data += 1);

    // More Locking!!

>     increment_in_mutex(&mut r);

Reviewer 2: `&mut` not needed here.

Reviewer 2: Sorry bad reading

The next thing is, **why should a lock be fallible**? One implementation of a spin-lock mutex was found which used a `Result` return to indicate a deadlock, but what other than a panic is correct if one would get a deadlock in a driver?

All these questions point to an infallible `Mutex` trait. The example of a deadlock in a driver is to simply implement this trait and `.unwrap()` in the implementation.

It boils down to this philosophy: `panic` is for program error - `Result` is for user error

#### Why is the trait `unsafe`?

The trait is marked `unsafe` in accordance with [RFC 19, 127], which states:

> Unsafe traits: An unsafe trait is a trait that is unsafe to implement, because it represents some kind of trusted assertion. Note that unsafe traits are perfectly safe to use. Send and Share are examples of unsafe traits: implementing these traits is effectively an assertion that your type is safe for threading.

> In this case the trusted assertion is that the data protect by the mutex is only accessed in a critical section, as we are going to share mutable references to the internal data. If this is not upheld by the implementation i.e. an unsound mutex implementation, the memory will be aliased mutably which is undefined behaviour. As the core requirement cannot be expressed using the Rust type-system in many cases, this requirement is placed on the implementor to uphold.

Reviewer 3: ```suggestion
In this case the trusted assertion is that the data protected by the mutex is only accessed in a critical section, as we are going to share mutable references to the internal data. If this is not upheld by the implementation i.e. an unsound mutex implementation, the memory will be aliased mutably which is undefined behaviour. As the core requirement cannot be expressed using the Rust type-system in many cases, this requirement is placed on the implementor to uphold.
```

> Another way to view this is that an unsound implementation of a mutex can lead to aliasing as the assumption made on the guard having `&mut self` is not as strict as one would expect from the Rust type-system.

Reviewer 2: I don't think it should be an unsafe trait as having this aliasing is unsound by definition, and thus no need do implement it unsafely.

# How We Teach This
[how-we-teach-this]: #how-we-teach-this

There will be 2 reference implementations after the acceptance of this RFC. One direct implementation based on critical sections in `cortex-m` and `risc-v` crates (that now reexport the [bare-metal mutex]) as the mutex will now be implemented in these crates and the reexport removed. And another in the [RTFM framework] which works together with procedural macros to give correct by construction resource handling and locks.

# Drawbacks
[drawbacks]: #drawbacks

As of writing this RFC I see no drawbacks.

# Alternatives
[alternatives]: #alternatives

The alternatives, taking `&self` and returning a `Result<...>` are discussed in [Detailed design](#detailed-design).

One can also further follow the [Rust standard library `Mutex`] and its RAII pattern, however this should not be pursued due to these causes:

1. The closure is explicit where it starts and ends, no need to wait for the destruction or add extra `{ ... }` around code to get predictable unlocks.
2. Makes code using locks more stable under refactoring, as small changes can drastically change the scope of a lock when using RAII.

> 3. A RAII based mutex can be `mem::forget` or `mem::swap` and all guarantees are out, the closure based mutex cannot be circumvented.

Reviewer 2: guarantees are not out, it will just not unlock, and thus deadlock if another lock is done, which is sound.

    * A RAII based mutex can be circumvented easily as follows:

```rust
// Break all guarantees
let x = lock1.lock();
let y = lock2.lock();
drop(x);

> drop(y);

Reviewer 2: I don't understand why it break all guarantees.

Reviewer 4: Imagine a lock that works in a "backup-and-restore" way for interrupts.

1. The first lock would backup the current Interrupt Flag (Active) and set it to Inactive.
2. The second lock would backup the current Interrupt Flag (Inactive) and set it to Inactive.
3. Unlocking the first lock would restore its saved Interrupt Flag (Active)
4. Unlocking the second lock would restore its saved Interrupt Flag (Inactive)

Of course it's possible to avoid this by implementing things differently (such as keeping a global depth count or something), but the closure-based design makes it much easier to avoid shooting yourself in the foot.

Reviewer 2: Then, that's more than RAII complexify the implementation of some mutex, or even avoid some implementation.

This can however easily be addressed by introducing a `Mutex` trait on which previous implementations can be built upon, which each architectural crate (cortex-m, risc-v, etc) reexports/implements, together with the [`embedded-hal`] which can use said trait in HAL design.

[bare-metal mutex]: https://github.com/japaric/bare-metal/blob/a2f8b222a14253141bbc7853e5699641e61e5ebf/src/lib.rs#L63
[RTFM mutex]: https://github.com/japaric/cortex-m-rtfm/blob/fafeeb27270ef24fc3852711c6032f65aa7dbcc0/src/lib.rs#L289
[mutex.rs]: https://gist.github.com/mvirkkunen/cc63d20648737022a8c5f93666093f4a
[`embedded-hal`]: https://github.com/rust-embedded/embedded-hal

# Detailed design
[detailed-design]: #detailed-design

## Trait

The trait is proposed as follows:

```rust
/// Any object implementing this trait guarantees exclusive access to the data contained
/// within the mutex for the duration of the lock.
pub trait Mutex {
    /// Data protected by the mutex
    type Data;

    /// Creates a critical section and grants temporary access to the protected data
    fn lock<R>(&mut self, f: impl FnOnce(&mut Self::Data) -> R) -> R;
}
```

> The design is such that the data protected by the mutex is only accessible within the closure provided to the `lock` method, moreover the closure shall run within a critical section such that it is guaranteed that no-one else can access the same data at the same time. The critical section can for example be:

Reviewer: ```suggestion
The design is such that the data protected by the mutex is only accessible within the closure provided to the `lock` method. Moreover, the closure shall run within a critical section such that it is guaranteed that no-one else can access the same data at the same time. The critical section can for example be:
```

* Disabling interrupts (e.g. `cortex-m` / `risc-v` critical sections)
* Analyzing the program and set BASEPRI in Cortex-M devices (Stack Resource Policy/RTFM way)
* Analyzing the program and set interrupt masks correctly
* etc

## Helpers

> A concerned that has been raised is that taking multiple locks can lead to excessive rightward drift as:

Reviewer: ```suggestion
A concern that has been raised is that taking multiple locks can lead to excessive rightward drift as:
```

                });
            });
        });
    });
});
```

It is proposed to provide a helper implementation to allow the following shorthand:

```rust
(a, b, c, d, e).lock(|a, b, c, d, e| {
    // ...
});
```

## Implementation

A proof-of-concept implementation of the crate with trait and helpers [can be found here](https://github.com/korken89/core-mutex).

### Design decisions and compatibility

####  The [Rust standard library `Mutex`] uses `&self` for the `lock` method, why choose `&mut self`?

This is the **most important** part of this RFC proposal. Let's assume that the implementation was `&self`, what would this mean? This would allow the following code to compile:

```rust
fn deadlock(mtx: &impl core_mutex::Mutex<Data = i32>) {
    mtx.lock(|data| {
        *data += 1;

        mtx.lock(|data| *data += 1);
    });
}
```

This example allows for a mutex to lock itself within a lock of itself, which is **deadlock by construction** - while if implementing it with `&mut self` the following would happen:

```rust
fn compile_error(mtx: &mut impl core_mutex::Mutex<Data = i32>) {
    mtx.lock(|data| {
        *data += 1;

        mtx.lock(|data| *data += 1);
    });
}
```

> Which produces the following error and making the mutex **deadlock-free by construction**:

Reviewer: ```suggestion
Which produces the following error and making the mutex **deadlock-safe by construction**:
```

The next thing is, **why should a lock be fallible**? One implementation of a spin-lock mutex was found which used a `Result` return to indicate a deadlock, but what other than a panic is correct if one would get a deadlock in a driver?

All these questions point to an infallible `Mutex` trait. The example of a deadlock in a driver is to simply implement this trait and `.unwrap()` in the implementation.

It boils down to this philosophy: `panic` is for program error - `Result` is for user error

#### Should the trait be marked `unsafe`?

> The trait does not need to be marked as `unsafe` as the signature of `lock` together with the Rust type system guarantees exclusive access. If there is an implementation which handles multiple access points to the same mutable reference `unsafe` code must be used, which puts the responsibility of upholding Rusts alias rule on the implementor.

Reviewer: ```suggestion
The trait does not need to be marked as `unsafe` as the signature of `lock` together with the Rust type system guarantees exclusive access. If there is an implementation which handles multiple access points to the same mutable reference, `unsafe` code must be used, which puts the responsibility of upholding Rust's alias rule on the implementor.
```

#### How does one implement generic functions that understand `lock`?

The following example shows how to write generic functions over the trait:

```rust
// A simple resource holding an `i32`
fn increment_in_mutex(mtx: &mut impl core_mutex::Mutex<Data = i32>) {
    mtx.lock(|data| *data += 1);
}
```

## FAQ

> Can this be used in a lock/unlock setting such as:

```rust
mtx.lock();

// Do something ...

mtx.unlock();
```

No. The `Mutex` trait is specifically for scoped use, i.e. the lock is defined in such a way that the resource protected by the lock is only accessible within the critical section that is the closure in the `lock` method.

This means that for example if you have a peripheral driver and you want to lock access to it while it does its work, this mutex is not designed for that. However you can implement a higher level mutex on top of this one.

> Should the `Mutex` trait not be in the [`embedded-hal`]?

It is our strong recommendation to **not** place it in the [`embedded-hal`] crate.

> If there are breaking changes in the [`embedded-hal`] it can be so that competing implementations of the Mutex trait starts to exist which could bifurcate the ecosystem based on version, plus that the Mutex trait will be tied to HAL releases.

Reviewer: ```suggestion
If there are breaking changes in the [`embedded-hal`] it can be so that competing implementations of the Mutex trait start to exist which could bifurcate the ecosystem based on version, plus that the Mutex trait will be tied to HAL releases.
```

# Drawbacks
[drawbacks]: #drawbacks

As of writing this RFC I see no drawbacks.

# Alternatives
[alternatives]: #alternatives

The alternatives, taking `&self` and returning a `Result<...>` are discussed in [Detailed design](#detailed-design).

One can also further follow the [Rust standard library `Mutex`] and its RAII pattern, however this should not be pursued due to these causes:

1. The closure is explicit where it starts and ends, no need to wait for the destruction or add extra `{ ... }` around code to get predictable unlocks.
2. Makes code using locks more stable under refactoring, as small changes can drastically change the scope of a lock when using RAII.
3. A RAII based mutex can be `mem::forget` or `mem::swap` and all guarantees are out, the closure based mutex cannot be circumvented.
    * A RAII based mutex can be circumvented easily as follows:

```rust
// Break all guarantees
let x = lock1.lock();
let y = lock2.lock();
drop(x);
drop(y);
```

# Unresolved questions
[unresolved]: #unresolved-questions

> * Should the trait be unsafe to indicate that one needs to be aware of the design constraints to implement it?

Reviewer: Can be removed now that this is settled?

Author: Yes, it should be removed - pushing that now

This can however easily be addressed by introducing a `Mutex` trait on which previous implementations can be built upon, which each architectural crate (cortex-m, risc-v, etc) reexports/implements, together with the [`embedded-hal`] which can use said trait in HAL design.

[bare-metal mutex]: https://github.com/japaric/bare-metal/blob/a2f8b222a14253141bbc7853e5699641e61e5ebf/src/lib.rs#L63
[RTFM mutex]: https://github.com/japaric/cortex-m-rtfm/blob/fafeeb27270ef24fc3852711c6032f65aa7dbcc0/src/lib.rs#L289
[mutex.rs]: https://gist.github.com/mvirkkunen/cc63d20648737022a8c5f93666093f4a
[`embedded-hal`]: https://github.com/rust-embedded/embedded-hal

# Detailed design
[detailed-design]: #detailed-design

## Trait

The trait is proposed as follows:

```rust
/// Any object implementing this trait guarantees exclusive access to the data contained
/// within the mutex for the duration of the lock.
pub trait Mutex {
    /// Data protected by the mutex
    type Data;

    /// Creates a critical section and grants temporary access to the protected data

>     fn lock<R>(&mut self, f: impl FnOnce(&mut Self::Data) -> R) -> R;

Reviewer 5: I must be missing something, but if you have a `&mut self`, why do you need a Mutex in the first place? In the `std::sync::Mutex` case, holding a `&mut self` allows to get a (wrapped) `&mut T` without a lock actually taking place.

Reviewer: Its explained in the this Readme's chapter named: **The [Rust standard library `Mutex`] uses `&self` for the `lock` method, why choose `&mut self`?**

Reviewer: Line 91

}
```

The design is such that the data protected by the mutex is only accessible within the closure provided to the `lock` method. Moreover, the closure shall run within a critical section such that it is guaranteed that no-one else can access the same data at the same time. The critical section can for example be:

* Disabling interrupts (e.g. `cortex-m` / `risc-v` critical sections)
* Analyzing the program and set BASEPRI in Cortex-M devices (Stack Resource Policy/RTFM way)
* Analyzing the program and set interrupt masks correctly
* etc

### Taxonomy

A "MUTual EXclusion" aka `Mutex` as used in this trait describes the **principle** and not the *implementation* characteristics. More specifically:

* A struct `Spinlock` could be one implementor of trait `Mutex` where the struct name gives away HOW mutual exclusion is achieved.
* A struct `Sleeplock` would be another implementor with a telling name.

> 		* For unfortunate reasons, a sleeping lock is often also called just `Mutex` in existing code (most prominent example being Linux). Using that naming scheme would mean a struct `Mutex` implementing the trait `Mutex` - this is discouraged to avoid stated ambiguity.

Reviewer: ```suggestion
  * For unfortunate reasons, a sleeping lock is often also called just `Mutex` in existing code (most prominent example being Linux). Using that naming scheme would mean a struct `Mutex` implementing the trait `Mutex` - this is discouraged to avoid stated ambiguity.
```

this still seems to render wrong.
